File: VERIS_Q2800_data_processing.m
%% VERIS DATA PROCESSING %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%  Ghent University, 2019
%  Daan Hanssens
%
%  Note: 0_fixed folder has to be included in path.
ccc;


%% USER-INPUT %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% File name (VERIS datafile)
filename = 'VSECmarie.DAT';

% Distance sensor - GPS
sensor_gps_dist = 1.10;  % m

% Coordinate system
coordinate_system = 'lam';  % lam : Lambert72, utm : UTM

% Grid size
dx = .25;  % m

% Save data as ascii and csv
save = 1;  % 1: save data, 0: don't save data


%% DATA %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Read data
file_id = fopen(strcat(filename)); % Grab file id
data_in = textscan(file_id, '%f %f %f %f %f'); % read
for ii = 1:length(data_in)
    data(:, ii) = data_in{ii}; % header: y, x, ec_25, ec_90, height
end

% Change coordinate system to UTM
if strcmp(coordinate_system, 'utm')
    [data(:, 1), data(:, 2)] = wgs2utm(data(:, 2), data(:, 1));
elseif strcmp(coordinate_system, 'lam')
    [data(:,1), data(:,2), data(:, 5)] = wgs2BD72(data(:, 1), ...
        data(:, 2), data(:, 5));
end

% Remove errors from VERIS data
data(data(:, 3) < 0, :) = [];
data(data(:, 4) < 0, :) = [];


%% SHIFT AND GRID DATA %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Shift tractrix
data(:, 1:2) = shift_tractrix(data(:, 1:2), sensor_gps_dist, 0, 0);

% Create mesh
s_mesh = create_blank_2D(data(:, 1:2), dx, 100, 1);

% Grid data
for ii = 3:length(data_in)
    
    % Gridding
    vqinterp = scatteredInterpolant(data(:,1), data(:,2), data(:,ii), ...
        'natural');
    vq_bl_h = vqinterp(s_mesh.rasX_bl, s_mesh.rasY_bl);
    vq = s_mesh.blank(:); vq(~isnan(vq)) = vq_bl_h;    
    
    % Create grid structure
    grids(:, :, ii-2) = reshape(vq, s_mesh.nrows, s_mesh.ncols);
    
end

% Save as ascii and csv
if save == 1
    save_ascii_raster_grid(strcat('VERIS_EC25.asc'), ...
        flipud(grids(:, :, 1)), s_mesh.ncols, s_mesh.nrows, ...
        s_mesh.minX, s_mesh.minY, dx, '%.2f ', s_mesh.nodatavalue);
    save_ascii_raster_grid(strcat('VERIS_EC90.asc'), ...
        flipud(grids(:, :, 2)), s_mesh.ncols, s_mesh.nrows, ...
        s_mesh.minX, s_mesh.minY, dx, '%.2f ', s_mesh.nodatavalue);
    dlmwrite('VERIS_data_corr.csv', data, 'precision', 10)
end


%% VISUALIZE %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Create title string
title_str = {'ECa ~25 cm (mS/m)', 'ECa ~90 cm (mS/m)'};

% Image plot
figure();
for ii = 1:2
    subplot(1, 2, ii);
    h = imagesc(s_mesh.xgv, s_mesh.ygv, grids(:, :, ii));
    set(h, 'AlphaData', ~isnan(grids(:, :, ii)));
    title(title_str(ii));
    colorbar();
    set(gca,'ydir','normal');
    axis equal; axis tight;
    xlabel('X (m)');
    ylabel('Y (m)');
end

% Contour plot
figure();
for ii = 1:2
    subplot(1, 2, ii);
    contourf(s_mesh.xgv, s_mesh.ygv, grids(:, :, ii), 10, ...
        'LineColor', 'none');
    title(title_str(ii));
    set(gca,'ydir','normal');
    colorbar();
    axis equal; axis tight;
    xlabel('X (m)');
    ylabel('Y (m)');
end

File: VERIS_Q2800_data_processing.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
	VERIS Q2800 DATA PROCESSING
	===========================
	ORBit, Ghent University, 2019
	Daan Hanssens

	Data-processing for VERIS Q2800 data.
	Note: ORBit module has to be included in path.
"""

# Import
import matplotlib.pyplot as plt
import ORBit
import pandas as pd
import utm


########################################################
# USER-INPUT
# ----------
#

# VERIS datafile
filename = 'data/VSECmarie.DAT'

# Distance sensor - GPS
sensor_gps_distance = 1.1  # (m)

# GPS antenna height
antenna_height = 1.45  # (m)

# Cell size and buffer for blank
cell_size = .2  # (m)
buffer = 2  # (m)

# Save data?
save_data = True  # True or False

# Coordinate system
change_wgs_to_utm = True


########################################################
# DATA
# ----
#

# Grab data
data = pd.read_csv(filename, sep='\t', header=None, names=['x', 'y', 'eca30', 'eca90', 'z'])

# Drop errors
data = data[~(data['eca30'] <= 0)]
data = data[~(data['eca90'] <= 0)]

# Calculate from WGS to UTM
if change_wgs_to_utm:
    for ii in range(data['x'].count()):
        data['x'].iloc[ii], data['y'].iloc[ii], _, _ = utm.from_latlon(
            data['y'].iloc[ii],
            data['x'].iloc[ii]
        )


########################################################
# SPATIAL CORRECTIONS
# -------------------
#

# Shift data
data.loc[:, ['x', 'y']] = ORBit.Spatial.shift_tractrix(data['x'].values, data['y'].values, sensor_gps_distance, 0, 0)

# Shift height
data.loc[:, ['z']] -= antenna_height

# Save to common
data.to_csv('common.csv')


########################################################
# INTERPOLATION
# -------------
#

# Create blank
blank = ORBit.Initialize.Blank(data['x'].values, data['y'].values, cell_size, buffer)

# Grid data
eca30_grd = ORBit.Spatial.natural_neighbor(data['x'].values, data['y'].values, data['eca30'].values, blank)
eca90_grd = ORBit.Spatial.natural_neighbor(data['x'].values, data['y'].values, data['eca90'].values, blank)


########################################################
# VISUALIZE
# ---------
#

# Visualize ECa 30
plt.figure()
plt.title('ECa 30')
plt.imshow(eca30_grd, origin='lower', extent=blank.extent, cmap='RdBu_r')
plt.xlabel('UTM X (m)'), plt.ylabel('UTM Y (m)')
plt.colorbar().set_label('ECa (mS/m)')
plt.clim(0, 50)
plt.axis('equal')

# Visualize ECa 90
plt.figure()
plt.title('ECa 90')
plt.imshow(eca90_grd, origin='lower', extent=blank.extent, cmap='RdBu_r')
plt.xlabel('UTM X (m)'), plt.ylabel('UTM Y (m)')
plt.colorbar().set_label('ECa (mS/m)')
plt.clim(0, 120)
plt.axis('equal')
ORBit.Other.save_figure('figs/eca90')